hello-world/src/lib.rs

```diff
@@ -6,74 +6,109 @@ static mut TAMAGOTCHI: Option<Tamagotchi> = None;
 
 #[derive(Default, Encode, Decode, TypeInfo)]
 pub struct Tamagotchi {
-   pub owner: ActorId,
    pub name: String,
    pub date_of_birth: u64,
+   pub owner: ActorId,
 
-   pub fed: u16,
-   pub happy: u16,
-   pub rested: u16,
+   pub fed: u64,
+   pub fed_block: u64,
+   pub entertained: u64,
+   pub entertained_block: u64,
+   pub rested: u64,
+   pub rested_block: u64,
 
-   pub last_ate: u32,
-   pub last_had_fun: u32,
-   pub last_slept: u32
+   pub allowed_account: Option<ActorId>,
 }
 
 impl Tamagotchi {
+    fn transfer(&mut self, actor_id: ActorId) {
+        let sender = msg::source();
+        assert!(
+            sender == self.owner || sender == self.allowed_account.unwrap_or_default(),
+            "Only owner or allowed account can transfer ownership"
+        );
+        self.owner = actor_id;
+
+        msg::reply(
+            TmgEvent::Transfer(actor_id), 
+            0
+        ).expect("Failed to share the TmgEvent");
+    }
+
+    fn approve(&mut self, actor_id: ActorId) {
+        assert!(msg::source() == self.owner, "Only owner can approve");
+        self.allowed_account = Some(actor_id);
+
+        msg::reply(
+            TmgEvent::Approve(actor_id),
+            0
+        ).expect("Failed to share the TmgEvent");
+    }
+
+    fn revoke_approval(&mut self) {
+        assert!(msg::source() == self.owner, "Only owner can revoke approval");
+        self.allowed_account = None;
+
+        msg::reply(
+            TmgEvent::RevokeApproval,
+            0
+        ).expect("Failed to share the TmgEvent");
+    }
+
     fn feed(&mut self) {
         assert!(msg::source() == self.owner, "Only owner can feed the tamagotchi");
         assert!(self.fed < 7000, "Tamagotchi not enough hungry");
 
         // calculating and normalizing hunger level
-        let hunger_level: u32 = (block_height() - self.last_ate) * HUNGER_PER_BLOCK;
-        let normalized_hunger_level: u16 = if hunger_level > MAX_FED {
+        let hunger_level: u64 = (block_height() as u64 - self.fed_block) * HUNGER_PER_BLOCK;
+        let normalized_hunger_level: u64 = if hunger_level > MAX_FED {
```

Comment:
Types can be inferred here:
```suggestion
        let hunger_level = (block_height() as u64 - self.fed_block) * HUNGER_PER_BLOCK;
        let normalized_hunger_level = if hunger_level > MAX_FED {
```